server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java

```diff
@@ -175,9 +177,15 @@ default void onStoreClosed(ShardId shardId) {}
      *
      * @param indexShard    the shard that is about to recover
      * @param indexSettings the shard's index settings
+     * @param notifiedSearchNodeIds list of search nodes where the shard is available
      * @param listener      listener notified when this step completes
      */
-    default void beforeIndexShardRecovery(IndexShard indexShard, IndexSettings indexSettings, ActionListener<Void> listener) {
+    default void beforeIndexShardRecovery(
+        IndexShard indexShard,
+        IndexSettings indexSettings,
+        Set<String> notifiedSearchNodeIds,
```

Comment:
IHMO, I don't like inserting `notifiedSearchNodeIds` here. It's a stateless parameter that somehow I feel we could handle in serverless code altogether, rather than hooking it up in this stateful interface.

I think we had a couple of ideas discussed, lets follow up on those again.